In [3]:
import math
import pandas as pd

## Decision tree (Entropy Function and Information Gain)

--------

## step 1 : 

# how i create node : 
#### - the attribute that best classifies the trining data  , use this attribute at the root od the tree 

# calculate { Entropy }

##### what the Entropy ..?

the Entropy is amount of the uncertainty in data set

#### You can find the law in Figure 1

# calculate { avg Entropy }

#### You can find the law in Figure 2

# calculate {  information gain } Difference  in entopy befor and after splitting data set in attribute 

#### You can find the law in Figure 3

----

#  step compute the entropy 

- compute the entropy for data set entropy (s)

- for evert arrtibute / feature  : 
   - claculte entropy for all onnther values entropy 
   - take avg information entropy for the currnt attribute 
   - calcute gain for the current attribute 

- pick the highest gain attribute 

- repete unitl we get the tree we desired 

example  : 



In [4]:
data = {
    
    "outlook": ["sunny", "sunny", "overcast", "rainy", "rainy", "rainy", "overcast", "sunny", "sunny", "rainy", "sunny", "overcast", "overcast", "rainy"],
    "humidity": ["high", "high", "high", "high", "normal", "normal", "normal", "high", "normal", "normal", "normal", "high", "normal", "high"],
    "temperature": ["hot", "hot", "hot", "mild", "cool", "cool", "cool", "mild", "cool", "mild", "mild", "mild", "hot", "mild"],
    "windy": ["weak", "strong", "weak", "weak", "weak", "strong", "strong", "weak", "weak", "weak", "strong", "strong", "weak", "strong"],
    "play_tennis": ["no", "no", "yes", "yes", "yes", "no", "yes", "no", "yes", "yes", "yes", "yes", "yes", "no"]
}





In [5]:
df = pd.DataFrame(data)


In [6]:
df

outlook humidity temperature   windy play_tennis
0      sunny     high         hot    weak          no
1      sunny     high         hot  strong          no
2   overcast     high         hot    weak         yes
3      rainy     high        mild    weak         yes
4      rainy   normal        cool    weak         yes
5      rainy   normal        cool  strong          no
6   overcast   normal        cool  strong         yes
7      sunny     high        mild    weak          no
8      sunny   normal        cool    weak         yes
9      rainy   normal        mild    weak         yes
10     sunny   normal        mild  strong         yes
11  overcast     high        mild  strong         yes
12  overcast   normal         hot    weak         yes
13     rainy     high        mild  strong          no

- the Possibility (yes)is ->> 9 in column play_tennis 

# and 

- the Possibility (no)is ->> 5 in column play_tennis 
------

- so that  total  is 14 


### law entorpy is (-p /p+n ).log2(p/p+n)-(n/p+n).log2(n/p+n)

In [7]:
p=0
n=0
for outcome in data["play_tennis"]:
    if outcome == "yes":
        p += 1
    elif outcome == "no":

        n += 1

In [8]:
p

9

In [9]:
n

5

In [10]:
p+n 

14

------------------------



In [11]:
def entropy(p, n):
    if p == 0 or n == 0:
        return 0
    else:
        return -((p/(p+n)) * math.log2(p/(p+n)) + (n/(p+n)) * math.log2(n/(p+n)))


In [12]:


def avg_entropy_for_all_features(df, target_column):
    feature_columns = df.columns[:-1] 
    avg_entropy_dict = {}

    for feature_column in feature_columns:
        unique_values = df[feature_column].unique()
        feature_avg_entropy = 0

        for value in unique_values:
            p = df[(df[feature_column] == value) & (df[target_column] == 'yes')].shape[0]
            n = df[(df[feature_column] == value) & (df[target_column] == 'no')].shape[0]

            value_entropy = entropy(p, n)
            feature_avg_entropy += (p + n) / (p + n) * value_entropy

        avg_entropy_dict[feature_column] = feature_avg_entropy / len(unique_values)

    return avg_entropy_dict


In [13]:
def gain(entropy, I):
    return (entropy - I)


--------------------


In [14]:
result = entropy(p , n)


In [15]:
result

0.9402859586706311

# the entropy(play_tennis) = 0.940

------

In [16]:
outlook_values = df["outlook"].unique()
outlook_values

array(['sunny', 'overcast', 'rainy'], dtype=object)

In [17]:
unique_values = df["outlook"].unique()
data_by_outlook = {}

for value in unique_values:
    p = df[(df["outlook"] == value) & (df["play_tennis"] == 'yes')].shape[0]
    n = df[(df["outlook"] == value) & (df["play_tennis"] == 'no')].shape[0]
    
    data_by_outlook[value] = {
        "Total Yes": p,
        "Total No": n
    }

for value, data in data_by_outlook.items():
    print(f"Outlook: {value}, {data}")
print("Total Yes for Sunny:", data_by_outlook["sunny"]["Total Yes"])
print("Total No for Rainy:", data_by_outlook["rainy"]["Total No"])


Outlook: sunny, {'Total Yes': 2, 'Total No': 3}
Outlook: overcast, {'Total Yes': 4, 'Total No': 0}
Outlook: rainy, {'Total Yes': 3, 'Total No': 2}
Total Yes for Sunny: 2
Total No for Rainy: 2


In [18]:
outlook_values = df["outlook"].unique()
entropy_results = {}

def calculate_entropy_for_category(df, category_column, target_column, category_value):
    p = df[(df[category_column] == category_value) & (df[target_column] == 'yes')].shape[0]
    n = df[(df[category_column] == category_value) & (df[target_column] == 'no')].shape[0]
    return entropy(p, n)

for value in outlook_values:
    result = calculate_entropy_for_category(df, "outlook", "play_tennis", value)
    entropy_results[value] = result

# طباعة النتائج للتحقق
for value, result in entropy_results.items():
    print(f"Entropy for {value}: {result}")

# أو يمكنك الوصول إلى النتائج باستخدام المفاتيح
print("Entropy for Sunny:", entropy_results["sunny"])
print("Entropy for Rainy:", entropy_results["rainy"])


Entropy for sunny: 0.9709505944546686
Entropy for overcast: 0
Entropy for rainy: 0.9709505944546686
Entropy for Sunny: 0.9709505944546686
Entropy for Rainy: 0.9709505944546686


# the entropy( outlook When the value is (sunny)) =0.9709505944546686
# the entropy( outlook When the value is (rainy)) =0.9709505944546686
# the entropy( outlook When the value is (overcast)) =0


-----

avg(p,n,x)=p+n / p+n * x

xsunny=Total Yes Sunny+Total No Sunny/Total Yes Sunny+Total No Sunny*Entropy Sunny

xrainy=Total Yes rainy+Total No rainy/Total Yes rainy+Total No rainy*Entropy rainy


# law is  : 

     p_sunny+n_sunny/p_sunny+n_sunny *entropy(outlook = sunny )
+
     p_overcast+n_overcast/p_overcast+n_overcast*entropy(outlook = overcast )
+
     p_rainy+n_rainy/p_rainy+n_rainy*entropy(outlook = rainy )
 


------

In [19]:
def avg(p, n, x):
    return (p + n) / (p + n) * x

x_sunny = avg(data_by_outlook["sunny"]["Total Yes"], data_by_outlook["sunny"]["Total No"], entropy_results["sunny"])
x_rainy = avg(data_by_outlook["rainy"]["Total Yes"], data_by_outlook["rainy"]["Total No"], entropy_results["rainy"])

result = x_sunny + x_rainy
print(result)


1.9419011889093372


In [20]:
x_r = avg(data_by_outlook["rainy"]["Total Yes"] ,data_by_outlook["rainy"]["Total No"] , entropy_results["rainy"])
x_r


0.9709505944546686

In [21]:
x_sunny+x_r

1.9419011889093372

----

-compute gain 
 - entropy(s) - I(attribute)

In [22]:

target_column = "play_tennis"


In [23]:

avg_entropy_dict = avg_entropy_for_all_features(df, target_column)
for feature, avg_entropy_value in avg_entropy_dict.items():
    print(f"{feature}:-: {avg_entropy_value}")


outlook:-: 0.6473003963031124
humidity:-: 0.7884504573082896
temperature:-: 0.9098579861712075
windy:-: 0.9056390622295665


In [24]:
total_p = df[df[target_column] == 'yes'].shape[0]
total_n = df[df[target_column] == 'no'].shape[0]
total_entropy = entropy(total_p, total_n)


In [25]:

avg_entropy_dict = avg_entropy_for_all_features(df, target_column)
gain_dict = {}


In [26]:

for feature, avg_entropy_value in avg_entropy_dict.items():
    gain_value = gain(total_entropy, avg_entropy_value)
    gain_dict[feature] = gain_value



In [27]:
print("Gain")
for feature, gain_value in gain_dict.items():
    print(f"{feature}: {gain_value}")

Gain
outlook: 0.2929855623675187
humidity: 0.15183550136234159
temperature: 0.03042797249942364
windy: 0.03464689644106467


-------------------------

## We will choose the largest value to put as the root in the decision tree because we are looking for the largest value present in gain in order to choose the most appropriate column.

-------------------------

the root node : 
  ## (outlook )

--------------------

and the next node  is : 
# (humidity)



---------------------

and the next node is : 
# (windy)

-----

and the next node is :
# (temperature)

-----

So, gain chooses the nodes according to the largest value to the smallest value that emerges from the data.